Tasks/Npm/npm.ts

```diff
@@ -0,0 +1,273 @@
+
+    let endpointId: string;
+    let feedId: string;
+    let registryLocation: string;
+    let npmRegistries: INpmRegistry[] = getLocalNpmRegistries(workingDir);
+    switch(command) {
+        case "install":
+            registryLocation = tl.getInput(NpmTaskInput.CustomRegistry);
+            endpointId = tl.getInput(NpmTaskInput.CustomEndpoint);
+            feedId = tl.getInput(NpmTaskInput.CustomFeed);
+            npm.line("install");
+            break;
+
+        case "publish":
+            registryLocation = tl.getInput(NpmTaskInput.PublishRegistry);
+            endpointId = tl.getInput(NpmTaskInput.PublishEndpoint);
+            feedId = tl.getInput(NpmTaskInput.PublishFeed);
+
+            // We delete their project .npmrc so it won't override our user .npmrc
+            tl.rmRF(`${workingDir}/.npmrc`);
+            npm.line("publish");
+            break;
+
+        case "custom":
+            registryLocation = tl.getInput(NpmTaskInput.CustomRegistry);
+            endpointId = tl.getInput(NpmTaskInput.CustomEndpoint);
+            feedId = tl.getInput(NpmTaskInput.CustomFeed);
+            npm.line(tl.getInput(NpmTaskInput.CustomCommand, true));
+            break;
+
+        default:
+            tl.setResult(tl.TaskResult.Failed, tl.loc("UnknownCommand", command));
+            return;
+    }
+
+    if (registryLocation == RegistryLocation.Feed) {
+        npmRegistries.push(NpmRegistry.FromFeedId(feedId));
+    } else if (endpointId) {
+        npmRegistries.push(NpmRegistry.FromServiceEndpoint(endpointId, command !== NpmCommand.Publish));
+    }
+
+    for (let registry of npmRegistries) {
+        if (registry.authOnly === false) {
+            appendToNpmrc(npmrc, `registry=${registry.url}\n`);
+        }
+
+        appendToNpmrc(npmrc, `${registry.auth}\n`);
+    }
+
+    await npm.exec();
+
+    tl.rmRF(npmrc);
+    popNpmrc(projectNpmrc);
+    tl.rmRF(getTempPath());
+}
+
+function getTempPath(): string {
+    let tempNpmrcDir
+        = tl.getVariable('Agent.BuildDirectory')
+        || tl.getVariable('Agent.ReleaseDirectory')
+        || process.cwd();
+    let tempPath = path.join(tempNpmrcDir, 'npm');
+    if (tl.exist(tempPath) === false) {
+        tl.mkdirP(tempPath);
+    }
+
+    return tempPath;
+}
+
+function getTempNpmrcPath(): string {
+    let tempNpmrcDir = getTempPath();
+    let tempUserNpmrcPath: string = path.join(
+        tempNpmrcDir,
+        `${tl.getVariable('Build.BuildId')}.npmrc`);
+    return tempUserNpmrcPath;
+}
+
+function appendToNpmrc(npmrc: string, data: string): void {
+    tl.writeFile(npmrc, data, {
+        flag: "a"
+    } as tl.FsOptions);
+}
+
+function pushNpmrc(npmrc: string): void {
+    if (npmrc) {
+        let tempPath = getTempPath();
+        let npmrcPath = path.join(tempPath, ".npmrc");
+
+        if (tl.exist(npmrc)) {
+            tl.debug(tl.loc("SavingProjectNpmrc"));
+            tl.cp(npmrc, path.join(tempPath, ".npmrc"), "-f");
```

Comment:
`getTempNpmrcPath` gets the full path to the user-level config file. In this case we want to preserve any project-level config files, so we store them in our temp path. This is necessary for the publish case, in which we want to override the project-level config, if any. I'll add some comments.